### DISCUSSION TITLE
Abort (Core dumped) when running cFS in docker container

### DISCUSSION THREAD

**Discussion by ivanperez-keera** [nasa/cFS#718] [Q&A] [answered]
Hi,

I'm trying to build cFS in docker, and I'm getting a core dump when I run it:

```Dockerfile
FROM i386/debian:bullseye

# Apt should not ask questions during configuration
ENV DEBIAN_FRONTEND=noninteractive

# Update packages available
RUN apt-get update

# cFS dependencies
RUN apt-get install -y cmake build-essential gcc-multilib g++-multilib

# Generic dependencies needed
RUN apt-get install -y git

# Get copy of cFS
RUN git clone --recursive https://github.com/nasa/cFS
WORKDIR cFS
RUN git submodule init
RUN git submodule update

RUN cp cfe/cmake/Makefile.sample Makefile
RUN cp -r cfe/cmake/sample_defs sample_defs

RUN make SIMULATION=native prep
RUN make
RUN make install
WORKDIR build/exe/cpu1/
CMD ./core-cpu1
```

Here's my session:
```
$ docker run -it cfs:latest 
CFE_PSP: Default Reset SubType = 1
CFE_PSP: Default CPU ID = 1
CFE_PSP: Default Spacecraft ID = 66
CFE_PSP: Default CPU Name: cpu1
OS_Posix_GetSchedulerParams():186:Policy 1: available, min-max: 1-99
OS_Posix_GetSchedulerParams():186:Policy 2: available, min-max: 1-99
OS_Posix_TaskAPI_Impl_Init():372:Selected policy 2 for RT tasks, root task = 99
OS_Posix_TaskAPI_Impl_Init():389:Could not setschedparam in main thread: Operation not permitted (1)
CFE_PSP: initializing module 'soft_timebase' with ID 0110ff00
CFE_PSP: Instantiated software timebase 'cFS-Master' running at 10000 usec
CFE_PSP: initializing module 'timebase_posix_clock' with ID 0110ff01
CFE_PSP: Using POSIX monotonic clock as CFE timebase
CFE_PSP: initializing module 'eeprom_mmap_file' with ID 0110ff02
CFE_PSP: Using MMAP simulated EEPROM implementation
CFE_PSP: EEPROM Range (2) created: Start Address = F7B2D000, Size = 00080000 Status = 0
CFE_PSP: initializing module 'ram_notimpl' with ID 0110ff03
CFE_PSP: Physical RAM access not implemented
CFE_PSP: initializing module 'port_notimpl' with ID 0110ff04
CFE_PSP: I/O Port access not implemented
CFE_PSP: initializing module 'iodriver' with ID 0110ff05
CFE_PSP: initializing module 'linux_sysmon' with ID 0110ff06
CFE_PSP: Starting the cFE with a POWER ON reset.
CFE_PSP: Clearing out CFE CDS Shared memory segment.
CFE_PSP: Clearing out CFE Reset Shared memory segment.
CFE_PSP: Clearing out CFE User Reserved Shared memory segment.
1980-002-17:04:50.28039 CFE_ES_SetupResetVariables: POWER ON RESET due to Power Cycle (Power Cycle).
1980-002-17:04:50.28041 CFE_ES_Main: CFE_ES_Main in EARLY_INIT state
CFE_PSP: CFE_PSP_AttachExceptions Called
1980-002-17:04:50.28312 CFE_ES_Main: CFE_ES_Main entering CORE_STARTUP state
1980-002-17:04:50.28313 CFE_ES_CreateObjects: Starting Object Creation calls.
1980-002-17:04:50.28314 CFE_ES_CreateObjects: Calling CFE_Config_Init
1980-002-17:04:50.28316 CFE_ES_CreateObjects: Calling CFE_ES_CDSEarlyInit
1980-002-17:04:50.28359 CFE_ES_CreateObjects: Calling CFE_EVS_EarlyInit
1980-002-17:04:50.28361 CFE_EVS_EarlyInit: Event Log cleared following power-on reset
1980-002-17:04:50.28361 CFE_ES_CreateObjects: Calling CFE_SB_EarlyInit
1980-002-17:04:50.28414 CFE_ES_CreateObjects: Calling CFE_TIME_EarlyInit
1980-012-14:03:20.00000 CFE_ES_CreateObjects: Calling CFE_TBL_EarlyInit
1980-012-14:03:20.00050 CFE_ES_CreateObjects: Calling CFE_FS_EarlyInit
OS_QueueCreate_Impl():119:OS_QueueCreate Error. errno = 22 (Invalid argument)
OS_QueueCreate_Impl():122:Your queue depth may be too large for the
OS_QueueCreate_Impl():123:OS to handle. Please check the msg_max
OS_QueueCreate_Impl():124:parameter located in /proc/sys/fs/mqueue/msg_max
OS_QueueCreate_Impl():125:on your Linux file system and raise it if you
OS_QueueCreate_Impl():126: need to or run as root
1980-012-14:03:20.05091 CFE_EVS_TaskInit: Call to CFE_SB_CreatePipe Failed:RC=0xCA000005
1980-012-14:03:20.05092 CFE_EVS_TaskMain: Application Init Failed,RC=0xCA000005
1980-012-14:03:20.05093 CFE_ES_ExitApp: CORE Application CFE_EVS Had an Init Error.
1980-012-14:03:20.05094 CFE_ES_ResetCFE: PROCESSOR RESET called from CFE_ES_ResetCFE (Commanded).
CFE_PSP: Exiting cFE with PROCESSOR Reset status.
Aborted (core dumped)
```

Does anyone know what's happening?

Thanks!

**Comment by ivanperez-keera** [comment]
For info: An alternative docker image, for Raspberry Pi, was also announced in https://github.com/nasa/cFS/discussions/399 by @astroesteban.

**Comment by ivanperez-keera** [comment, marked as answer]
TL;DR: The problem in my case was that the `msg_queue` size was lower than required.

# Details

To anybody who encounters the same problem, here's what I did to figure out the solution.

The bottom of the log indicates what the error may be:
```
1980-012-14:03:20.05091 CFE_EVS_TaskInit: Call to CFE_SB_CreatePipe Failed:RC=0xCA000005
1980-012-14:03:20.05092 CFE_EVS_TaskMain: Application Init Failed,RC=0xCA000005
1980-012-14:03:20.05093 CFE_ES_ExitApp: CORE Application CFE_EVS Had an Init Error.
1980-012-14:03:20.05094 CFE_ES_ResetCFE: PROCESSOR RESET called from CFE_ES_ResetCFE (Commanded).
CFE_PSP: Exiting cFE with PROCESSOR Reset status.
Aborted (core dumped)
```

However, it's not necessarily clear that the abort (core dumped) was caused by the cFS stopping. I first compiled cFS with debug enabled. Change the `make prep` line to:

```
RUN make SIMULATION=native BUILDTYPE=debug prep
```

Then I installed `gdb` and `vim` inside the Docker container:
```
apt-get install --yes gdb vim
```

I then run cFS with `gdb`. The output is too long, the interesting bit is that it crashes, and one of the apps outputs the error code 0xCA000005.

<details open>
  <summary>Full output</summary>

  ```
  # gdb ./core-cpu1
  GNU gdb (Debian 13.1-3) 13.1
  Copyright (C) 2023 Free Software Foundation, Inc.
  License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
  This is free software: you are free to change and redistribute it.
  There is NO WARRANTY, to the extent permitted by law.
  Type "show copying" and "show warranty" for details.
  This GDB was configured as "x86_64-linux-gnu".
  Type "show configuration" for configuration details.
  For bug reporting instructions, please see:
  <https://www.gnu.org/software/gdb/bugs/>.
  Find the GDB manual and other documentation resources online at:
      <http://www.gnu.org/software/gdb/documentation/>.

  For help, type "help".
  Type "apropos word" to search for commands related to "word"...
  Reading symbols from ./core-cpu1...
  (gdb) run
  Starting program: /cFS/build/exe/cpu1/core-cpu1
  warning: Error disabling address space randomization: Operation not permitted
  [Thread debugging using libthread_db enabled]
  Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
  CFE_PSP: Default Reset SubType = 1
  CFE_PSP: Default CPU ID = 1
  CFE_PSP: Default Spacecraft ID = 66
  CFE_PSP: Default CPU Name: cpu1
  OS_Posix_GetSchedulerParams():186:Policy 1: available, min-max: 1-99
  OS_Posix_GetSchedulerParams():186:Policy 2: available, min-max: 1-99
  OS_Posix_TaskAPI_Impl_Init():372:Selected policy 2 for RT tasks, root task = 99
  OS_Posix_TaskAPI_Impl_Init():389:Could not setschedparam in main thread: Operation not permitted (1)
  [New Thread 0x7fd852fdc6c0 (LWP 16)]
  CFE_PSP: initializing module 'soft_timebase' with ID 0110ff00
  [New Thread 0x7fd852cb16c0 (LWP 17)]
  CFE_PSP: Instantiated software timebase 'cFS-Master' running at 10000 usec
  CFE_PSP: initializing module 'timebase_posix_clock' with ID 0110ff01
  CFE_PSP: Using POSIX monotonic clock as CFE timebase
  CFE_PSP: initializing module 'eeprom_mmap_file' with ID 0110ff02
  CFE_PSP: Using MMAP simulated EEPROM implementation
  CFE_PSP: initializing module 'ram_notimpl' with ID 0110ff03
  CFE_PSP: Physical RAM access not implemented
  CFE_PSP: initializing module 'port_notimpl' with ID 0110ff04
  CFE_PSP: I/O Port access not implemented
  CFE_PSP: initializing module 'iodriver' with ID 0110ff05
  CFE_PSP: initializing module 'linux_sysmon' with ID 0110ff06
  CFE_PSP: EEPROM Range (2) created: Start Address = 7FD852C2D000, Size = 00080000 Status = 0
  CFE_PSP: Starting the cFE with a POWER ON reset.
  CFE_PSP: Clearing out CFE CDS Shared memory segment.
  CFE_PSP: Clearing out CFE Reset Shared memory segment.
  CFE_PSP: Clearing out CFE User Reserved Shared memory segment.
  1980-001-13:36:47.51609 CFE_ES_SetupResetVariables: POWER ON RESET due to Power Cycle (Power Cycle).
  1980-001-13:36:47.51614 CFE_ES_Main: CFE_ES_Main in EARLY_INIT state
  CFE_PSP: CFE_PSP_AttachExceptions Called
  1980-001-13:36:47.51957 CFE_ES_Main: CFE_ES_Main entering CORE_STARTUP state
  1980-001-13:36:47.51957 CFE_ES_CreateObjects: Starting Object Creation calls.
  1980-001-13:36:47.51958 CFE_ES_CreateObjects: Calling CFE_Config_Init
  1980-001-13:36:47.51960 CFE_ES_CreateObjects: Calling CFE_ES_CDSEarlyInit
  1980-001-13:36:47.51990 CFE_ES_CreateObjects: Calling CFE_EVS_EarlyInit
  1980-001-13:36:47.51992 CFE_EVS_EarlyInit: Event Log cleared following power-on reset
  1980-001-13:36:47.51992 CFE_ES_CreateObjects: Calling CFE_SB_EarlyInit
  1980-001-13:36:47.52026 CFE_ES_CreateObjects: Calling CFE_TIME_EarlyInit
  1980-012-14:03:20.00000 CFE_ES_CreateObjects: Calling CFE_TBL_EarlyInit
  1980-012-14:03:20.00040 CFE_ES_CreateObjects: Calling CFE_FS_EarlyInit
  [New Thread 0x7fd852c2c6c0 (LWP 18)]
  OS_QueueCreate_Impl():119:OS_QueueCreate Error. errno = 22 (Invalid argument)
  OS_QueueCreate_Impl():122:Your queue depth may be too large for the
  OS_QueueCreate_Impl():123:OS to handle. Please check the msg_max
  OS_QueueCreate_Impl():124:parameter located in /proc/sys/fs/mqueue/msg_max
  OS_QueueCreate_Impl():125:on your Linux file system and raise it if you
  OS_QueueCreate_Impl():126: need to or run as root
  1980-012-14:03:20.05097 CFE_EVS_TaskInit: Call to CFE_SB_CreatePipe Failed:RC=0xCA000005
  1980-012-14:03:20.05099 CFE_EVS_TaskMain: Application Init Failed,RC=0xCA000005
  1980-012-14:03:20.05099 CFE_ES_ExitApp: CORE Application CFE_EVS Had an Init Error.
  1980-012-14:03:20.05100 CFE_ES_ResetCFE: PROCESSOR RESET called from CFE_ES_ResetCFE (Commanded).
  CFE_PSP: Exiting cFE with PROCESSOR Reset status.

  Thread 4 "CFE_EVS" received signal SIGABRT, Aborted.
  [Switching to Thread 0x7fd852c2c6c0 (LWP 18)]
  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
  44	./nptl/pthread_kill.c: No such file or directory.
  ```
</details>

I then run the `gdb` command `backtrace` to figure out what had caused the core dump:
```
(gdb) backtrace
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
#1  0x00007fd852e80d9f in __pthread_kill_internal (signo=6, threadid=<optimized out>) at ./nptl/pthread_kill.c:78
#2  0x00007fd852e31f32 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#3  0x00007fd852e1c472 in __GI_abort () at ./stdlib/abort.c:79
#4  0x000055a7249c693e in CFE_PSP_Restart (reset_type=1) at /cFS/psp/fsw/pc-linux/src/cfe_psp_support.c:113
#5  0x000055a72498ff20 in CFE_ES_ResetCFE (ResetType=1) at /cFS/cfe/modules/es/fsw/src/cfe_es_api.c:113
#6  0x000055a724990564 in CFE_ES_ExitApp (ExitStatus=8) at /cFS/cfe/modules/es/fsw/src/cfe_es_api.c:399
#7  0x000055a72499ec2e in CFE_EVS_TaskMain () at /cFS/cfe/modules/evs/fsw/src/cfe_evs_task.c:213
#8  0x000055a72499307d in CFE_ES_TaskEntryPoint () at /cFS/cfe/modules/es/fsw/src/cfe_es_apps.c:566
#9  0x000055a7249bbde9 in OS_TaskEntryPoint (task_id=65537) at /cFS/osal/src/os/shared/src/osapi-task.c:132
#10 0x000055a7249c05ab in OS_PthreadTaskEntry (arg=0x10001) at /cFS/osal/src/os/posix/src/os-impl-tasks.c:119
#11 0x00007fd852e7f044 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#12 0x00007fd852efe880 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:100
```

The message pointed me to the `CFE_ES` and `CFE_EVS` modules. Opening those showed that the app was _voluntarily_ closing, and there's an `abort` call at the end that causes it to finish... not gracefully (in the unix process sense).

Tracking that back took me to a call to the pipe initialization, which was returning the error code I pointed at above.

Searching for that error code in the tree brought up its definition. I had to use case-insensitive search.

```
# grep -niIHre 'CA000005' /cFS/
/cFS/cfe/modules/core_api/fsw/inc/cfe_error.h:772:#define CFE_SB_PIPE_CR_ERR ((CFE_Status_t)0xca000005)
```

Opening that location `/cFS/cfe/modules/core_api/fsw/inc/cfe_error.h` brought up a description of the error:

https://github.com/nasa/cFE/blob/b72dd4e1f9f44c7dbb7a12895b5ac1635eb239b2/modules/core_api/fsw/inc/cfe_error.h#L763-L772

That pointed me to the max queues issue. When I investigated further, I saw that the message queue size was too small:

```
# cat /proc/sys/fs/mqueue/msg_max
10
```

I knew from prior experiments that this could be an issue, so I tried to increase it by running docker with a sysctl configuration setting:

```
$ docker run --sysctl fs.mqueue.msg_max=2048 -it cfs:latest /bin/bash
```

That solve the issue for me.

I also looked at increasing the message queue limit with the `docker run` argument `--ulimit msgqueue=-1` but that wasn't necessary in the end.

**Reply by ivanperez-keera** [reply]
For environments where modifying `fs.mqueue.msg_max` is not possible, like rootless docker, one option is to run with a limited queue size (e.g., whatever the container is running with). However, in such cases, if running as root, cfs (specifically osal) will ignore the OS's `msg_max`, which will lead to the same crash described originally here. A possible solution is to make osal not check the user ID (see: https://github.com/nasa/osal/issues/73#issuecomment-1840961452). An alternative that could also work would be to install everything with a different user (not root).

**Comment by ivanperez-keera** [comment]
This requires privileged thought, which are not available for many users in environments where cFS may be used due to security policies.

I remember at some point hearing about the use of mqueues possibly being replaced in cFS. I wonder what the status of that is.

**Reply by skliper** [reply]
I see you already found the open issue at https://github.com/nasa/osal/issues/73, referencing here for others.
